src/native/libs/System.Security.Cryptography.Native/pal_evp_pkey_ml_dsa.c

```diff
@@ -0,0 +1,297 @@
+#include <assert.h>
+
+// ML-DSA-44 signature length is 2420 bytes.
+#define SHORTEST_POSSIBLE_ML_DSA_SIG_LEN 2420
+
+int32_t CryptoNative_IsSignatureAlgorithmAvailable(const char* algorithm)
+{
+#ifdef FEATURE_DISTRO_AGNOSTIC_SSL
+    if (!API_EXISTS(EVP_PKEY_sign_message_init) ||
+        !API_EXISTS(EVP_PKEY_verify_message_init))
+    {
+        return 0;
+    }
+#endif
+
+    int32_t ret = 0;
+#ifdef HAVE_OPENSSL_EVP_PKEY_SIGN_MESSAGE_INIT
+    EVP_SIGNATURE* sigAlg = NULL;
+
+    sigAlg = EVP_SIGNATURE_fetch(NULL, algorithm, NULL);
+    if (sigAlg)
+    {
+        ret = 1;
+        EVP_SIGNATURE_free(sigAlg);
+    } 
+#endif
+
+    (void)algorithm;
+    return ret;
+}
+
+EVP_PKEY* CryptoNative_MLDsaGenerateKey(const char* keyType, uint8_t* seed, int32_t seedLen)
+{
+#ifdef FEATURE_DISTRO_AGNOSTIC_SSL
+    if (!API_EXISTS(EVP_PKEY_sign_message_init) ||
+        !API_EXISTS(EVP_PKEY_verify_message_init))
+    {
+        return NULL;
+    }
+#endif
+
+#ifdef HAVE_OPENSSL_EVP_PKEY_SIGN_MESSAGE_INIT
+    ERR_clear_error();
+
+    if (seed && seedLen != 32)
+    {
+        return NULL;
+    }
+
+    EVP_PKEY_CTX* pctx = EVP_PKEY_CTX_new_from_name(NULL, keyType, NULL);
+    EVP_PKEY* pkey = NULL;
+
+    if (!pctx)
+    {
+        return NULL;
+    }
+
+    if (EVP_PKEY_keygen_init(pctx) <= 0)
+    {
+        goto done;
+    }
+
+    if (seed)
+    {
+        OSSL_PARAM params[2];
+        params[0] = OSSL_PARAM_construct_octet_string(OSSL_PKEY_PARAM_ML_DSA_SEED, (void*)seed, Int32ToSizeT(seedLen));
+        params[1] = OSSL_PARAM_construct_end();
+
+        if (EVP_PKEY_CTX_set_params(pctx, params) <= 0)
+        {
+            goto done;
+        }
+    }
+
+    if (EVP_PKEY_keygen(pctx, &pkey) != 1 && pkey != NULL)
+    {
+        EVP_PKEY_free(pkey);
+        pkey = NULL;
+    }
+
+done:
+    if (pctx != NULL)
+    {
+        EVP_PKEY_CTX_free(pctx);
+    }
+
+    return pkey;
+#else
+    (void)keyType;
+    (void)seed;
+    (void)seedLen;
+    return NULL;
+#endif
+}
+
+static bool IsValidMLDsaAlgorithm(const char* alg)
+{
+    return alg != NULL && (strcmp(alg, "ML-DSA-44") == 0 ||
+                           strcmp(alg, "ML-DSA-65") == 0 ||
+                           strcmp(alg, "ML-DSA-87") == 0);
+}
+
+int32_t CryptoNative_MLDsaSignPure(EVP_PKEY *pkey,
+                                   void* extraHandle,
+                                   uint8_t* msg, int32_t msgLen,
+                                   uint8_t* context, int32_t contextLen,
+                                   uint8_t* destination, int32_t destinationLen)
+{
+#ifdef FEATURE_DISTRO_AGNOSTIC_SSL
+    if (!API_EXISTS(EVP_PKEY_sign_message_init) ||
+        !API_EXISTS(EVP_PKEY_verify_message_init))
+    {
+        return -1;
+    }
+#endif
+
+#ifdef HAVE_OPENSSL_EVP_PKEY_SIGN_MESSAGE_INIT
+    ERR_clear_error();
+
+    if (!pkey || !msg || !destination || msgLen < 0 || contextLen < 0 || destinationLen < SHORTEST_POSSIBLE_ML_DSA_SIG_LEN)
+    {
+        return -1;
+    }
```

Comment:
Algorithms that we had during .NET Core 1.0 tend to use this pattern, because I was concerned that the shim DLL might leak out as callable API (particularly since it was distributed via NuGet).

Now that we have a pretty firm "NO" on that, we generally `assert()` these conditions instead.  We're the only caller, we should never get this wrong.

Still do the assertion, because tearing down the process from an assert is OK, OOB-writes is not.